js/XAPI.js

```diff
@@ -1125,6 +1124,10 @@ class XAPI extends Backbone.Model {
     });
 
     // Pass the new state to the LRS.
+    this.sendStateToServer(activityId, actor, collectionName, registration, newState);
+  }
+
+  sendStateToServer(activityId, actor, collectionName, registration, newState) {
```

Comment:
I don't think `activityId`, `actor` and `registration` need to be passed, as they come straight from:

```js
    const activityId = this.get('activityId');
    const actor = this.get('actor');
    const registration = this.get('shouldUseRegistration') === true
      ? this.get('registration')
      : null;
```

The values for `collectionName` and `newState` come from the following parts of Adapt.

The XAPI plugin uses [sendState](https://github.com/adaptlearning/adapt-contrib-xapi/blob/267ecbb3c0bbed395a8284412bb59c9fca351e50/js/setupOfflineStorage.js#L51) and a [fakeModel](https://github.com/adaptlearning/adapt-contrib-xapi/blob/267ecbb3c0bbed395a8284412bb59c9fca351e50/js/setupOfflineStorage.js#L7-L11) as the argument `model` to save offlineStorage stuff.

The core sends actual models with their corresponding state object from [adaptModel.js](https://github.com/adaptlearning/adapt-contrib-core/blob/a1b6c57f8608d0a38ef6918ce3bc9bad6961f580/js/models/adaptModel.js#L235-L239)

The `collectionName` value is derived from the `_type` of the `model` attribute (offlineStorage, component, block, course, menu, page, article) and its corresponding xapi type here:
```js
    const type = model.get('_type');
    const collectionName = _.findKey(this.coreObjects, o => {
      return (o === type || o.indexOf(type) > -1);
    });
```
https://github.com/adaptlearning/adapt-contrib-xapi/blob/267ecbb3c0bbed395a8284412bb59c9fca351e50/js/XAPI.js#L60-L68

The `newState` sent to the server is either a collection of block, article, component or contentObject states or the course or offlineStorage state:
https://github.com/adaptlearning/adapt-contrib-xapi/blob/267ecbb3c0bbed395a8284412bb59c9fca351e50/js/XAPI.js#L1105-L1125

What this means is that debounce cannot happen on the `this.sendStateToServer` function unless the function is passed a list of the changed `collectionNames` since the last call to `this.sendStateToServer`. With a list of changed `collectionNames` it can look up the current state objects from `this.get('state')` and then send to them to the server as a batch of `this.xapiWrapper.sendState` requests.

i.e.

`sendState` should update `this.set({ state })` as it does currently but should also update a list like `this.changedCollectionNames[collectionName] = true;` before it calls the debounced `this.sendStateToServer`. 

`sendStateToServer` should then iterate through the changed collection names, fetching the relevant state from the state object and use both of those variables plus the aforementioned variables (activityId, etc) to send a series of `this.xapiWrapper.sendState` requests.

```js
    const activityId = this.get('activityId');
    const actor = this.get('actor');
    const registration = this.get('shouldUseRegistration') === true
      ? this.get('registration')
      : null;
    const changedCollectionNames = Object.keys(this.changedCollectionNames) // capture only the names
    this.changedCollectionNames = {} // reset for the next round
    for (const collectionName of changeCollectionNames) {
       const newState = this.get('state')[collectionName];
       await new Promise(resolve => {
  this.xapiWrapper.sendState(activityId, actor, collectionName, registration, newState, null, null, (error, xhr) => {
           if (error) {
             Adapt.trigger('xapi:lrs:sendState:error', error);
             return resolve(); // as the error is handled by event we shouldn't send it back in the promise
           }
           Adapt.trigger('xapi:lrs:sendState:success', newState);
           return resolve();
        });
      });
    }
```
The function should not be passed